### PR TITLE
WIP: Unwrapping UnionAll types for datatype definition in REPL help

### PR DESCRIPTION
**Pull request by m-j-w**
This addresses  #21029. The proposed solution has served me well today, so maybe it's worth a PR.
I added tests. Still, I'm unsure whether this solves the issue completely.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
This is probably basically the right thing to do, but note that the result of `unwrap` is not a valid type, so for example the `subtypes` call might not work.

**Comment by m-j-w** [comment]
How would I solve this potential subtype printing issue? I didn't observe wrong printing of subtypes of abstract types. Neither with the same or additional type parameters.
```jl
julia> abstract type a{T<:Number,N} end
       primitive type p{T<:Number,N,A}<:a{T,N} 32 end
       struct s{T<:Number,N} <: a{T,N}
          a::NTuple{N,T}
       end

help?> a
search: any all abs Any ANY atan asin asec any! all! airy acsc acot acos abs2 atanh atand atan2 asinh asind asech asecd ascii angle airyx acsch acscd acoth

  No documentation found.

  Summary:

  abstract type a{T<:Number,N} <: Any

  Subtypes:

  p{T<:Number,N,A} where A
  s{T<:Number,N}

  Note, a is represented by a wrapping type UnionAll that describes the bounds of its type parameters.
```

**Comment by JeffBezanson** [comment]
Yes, it might happen to work in a number of cases. See the `subtypes` function itself for a good example --- it keeps both a wrapped and unwrapped version of a type, and uses the unwrapped version when it needs to look at the structure and the wrapped version when it needs to call `typeintersect`. That's generally a good pattern to use.

**Comment by m-j-w** [comment]
I see. I'll try to work that out.

**Comment by m-j-w** [comment]
Accidental push and still WIP. (I'm aware of the unnecessary additional scopes...)

**Comment by m-j-w** [comment]
@JeffBezanson would you mind giving 11a26c1 a quick look ?
To me it seems `UnionAll` are printed correctly. I also gave `Union` a try which isn't yet handled either, but a `UnionAll` of a `Union` is kind of tricky; and I'm not really sure I understand the overall approach taken e.g. in `show(IOContext(io, :unionall_env => x.var), x.body)`. Thanks !

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -373,28 +373,31 @@ function summarize(io::IO, λ::Function, binding)
...
 
-function summarize(io::IO, U::UnionAll, binding)
-    summarize(io, Base.unwrap_unionall(U), binding)
-    println(io, "Note, ```", Base.typename(U),
-            "``` is represented by a wrapping type ```UnionAll``` that " *
-            "describes the bounds of its type parameters.\n")
+function summarize(io::IO, T::Type{Union}, binding)
+    println(io, "**Types:**")
+    t = getfield(binding.mod, binding.var)
+    for ut in Core.Inference.uniontypes(Base.unwrap_unionall(t))
```
Comment (JeffBezanson) [inline base/docs/Docs.jl]: `uniontypes` also exists in Base.

File: base/docs/Docs.jl [hunk 2]
```diff
@@ -373,28 +373,31 @@ function summarize(io::IO, λ::Function, binding)
...
+        print(io, " - `")
+        Base.show_datatype(io, isa(ut, UnionAll) ? Base.unwrap_unionall(ut) : ut)
```
Comment (JeffBezanson) [inline base/docs/Docs.jl]: I don't really understand the purpose of this code. We already have `show` methods for types; why not just call `show`?

**Pull request closed without merging**